platform/m6/sub/101a/boot.c

```diff
@@ -0,0 +1,333 @@
+#include "cache.h"
+#include "core.h"
+#include "lolevel.h"
+
+#include "../../../generic/mmu_utils.h"
+#include "boot.h"
+
+#define CHDK_MMUTBL_START 0xdffda400
+#define CHDK_PAGES_START  0x11d0000
+#define CANON_TABLES_AT   0xdffc0000
+
+static int no_chdk_please = 0;
+
+void CreateTask_spytask() {
+	if (!no_chdk_please) {
+		_CreateTask("SpyTask", 0x19, 0x2000, core_spytask, 0);
+	}
+}
+
+void startup_mode_fix(void) {
+	int startupmode = *(int*) 0x9eb0;
+	switch (startupmode) {
+	case 0x80000:
+	case 0x400000:
+	case 0x200000:
+		no_chdk_please = 1;
+		break;
+	}
+}
+
+void __attribute__((naked,noinline,aligned(4))) patch_E0020668() {
+	asm volatile (
+			"bl         CreateTask_spytask\n"
+			"bl         sub_E02796AE\n" // execute overwritten instruction
+			"bl         sub_E010E954\n"// execute overwritten instruction
+			"ldr        pc, =0xe0020671\n"// -> back to ROM
+	);
+}
+
+void plant_hacks_for_e0020000(unsigned addr) {
+	// switch off card instead of exec'ing diskboot
+	*(unsigned*) (addr + 0x0644) = BL_INSTR_T1(0xE0020644, 0xE04B3F3A);
+	// conditionally start SpyTask
+	place_fw_patch_t2_64b(patch_E0020668, addr + 0x0668);
+}
+
+void __attribute__((naked,noinline,aligned(4))) patch_E0079DA4() {
+	asm volatile (
+			"bl         sub_e007006e\n" // startup mode inquiry
+			"bl         startup_mode_fix\n"// +
+			"bl         sub_e007a49c\n"// <kbd_read_keys>
+			"ldr        r3, =0x6a1a8\n"
+			"ldr        pc, =0xe0079dad\n"// -> back to ROM
+	);
+}
+
+void __attribute__((naked,noinline,aligned(4))) patch_E007A1E8() {
+	asm volatile (
+			"ldr        r3, =mykbd_task\n"
+			"movs       r0, #0\n"
+			"mov        r2, #0x2000\n"
+			"strd       r0, r1, [sp]\n"
+			"ldr        pc, =0xe007a1f1\n" // -> back to ROM
+	);
+}
+
+void plant_hacks_for_e0070000(unsigned addr) {
+	// insert startup mode handling
+	place_fw_patch_t2_64b(patch_E0079DA4, addr + 0x9DA4);
+	place_fw_patch_t2_64b(patch_E007A1E8, addr + 0xA1E8);
+	// nop (benign 16-bit instruction, added to replace 1st half of destroyed 32-bit instruction)
+	*(unsigned short*) (addr + 0xA1E6) = 0xbf00;
+}
```

Comment:
@srsa-hu: added a patch for the keyboard task but I'm not sure about the NOP assignment on line 72.

I've looked at your various patching for the M100 and I don't fully understand why you are only doing this operation for the last patched address ?

I've used **0xe007A1E6** which is the address before the patch but it looks like the camera is not booting anymore...
Initially you've assigned the address **0xe0079DA2** which was the address above the 1st patch.

Reply:
here is the Ghidra section -> https://pasteboard.co/PS73xUsHb1bo.png

Reply:
for the **plant_hacks_for_e0020000** function; you haven't specified a NOP assignment so I'm not really 100% when it's needed...

Reply:
ok I've swapped line 69 & 70 and assigned the previous address (0xe0079DA2); I've got now the ALT mode from the playback button. I can navigate in the CHDK menus...

It'll be nice to understand how to patch properly...

Reply:
not sure it's stable, by just navigating the menus -> crash!

Reply:
The version of mmuutils you have is missing a few more explanations.
The functions place_fw_hook_t2_64b() and place_fw_patch_t2_64b() both clobber 2 words (8 bytes) at 'origram'.
'origram' is required to be word aligned (must be multiples of 4).
_If_, and only if 'origram' points in the inside of a 32-bit Thumb instruction, then you need to neutralize the first 2 bytes of that instruction, that's what happens when I write 0xbf00 to a 2-aligned address. 
The goal of your patch is to modify what's placed in registers R3 (task's function pointer) and  R2 (stack size). All other  firmware code must retain its original functionality.
Your patch seems correct. Patching +0xa1e6 is not necessary as you modify R3 in your patch anyway.

In [this commit](https://github.com/cedricbompart/chdk/commit/9e7ba04df00b761930c52fba23c30e74bca27b1b) you removed an important line from plant_hacks_for_e0070000() that patched the remains of a 32-bit instruction - that was probably the reason why it did not boot. The line you removed was part of the existing patch.

When it's crashing, getting a romlog can help.

Reply:
thanks